src/psyclone/psyir/symbols/symboltable.py

```diff
@@ -83,6 +83,39 @@ def __init__(self, node=None):
                 "Schedule or a Container but found '{0}'."
                 "".format(type(node).__name__))
         self._node = node
+        # The default visibility of symbols in this symbol table. (Required
+        # when we have symbols of UnknownType since their associated
+        # declaration may or may not contain visibility information.)
+        self._default_visibility = None
```

Comment:
I would prefer that we gave a default value to it (the same that Fortran does when nothing is specified). Instead of having the possibility for it to be None, because then we technically check that is "not None" every time we use it.

Document the default value in the docstring

Reply:
Yes, that is better. However, that made me revisit what I'd done in `process_declarations()` and I decided that the current implementation (which had side effects for the symbol table of the supplied PSyIR node) wasn't very nice. I've therefore removed the call to `_parse_access_statements` from inside it and added it, where necessary, before every location in the code where `process_declarations()` is called. Due to this 'promotion' I've also renamed `_parse_access_statements` to `process_access_statements`.

Reply:
I like how this turned out, having the `process_access_statements` as a previous step to `process_declarations` makes things more clear in my view.

For the symbol table constructor, since this is not setting a mutable container, you could encode the default value directly into the function definition: ` def __init__(self, node=None, default_visibility=Symbol.Visibility.PUBLIC):`. I think it makes for a bit better documentation.
And for the docstring of this parameter I suggest something like: "optional default visibility value for this symbol table, if not provided it defaults to PUBLIC visibility."

Reply:
Oh yes, that's nicer - thanks.